apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -308,14 +293,22 @@ choose(N, List) ->
     List2 = [X || {_, X} <- List1],
     lists:sublist(List2, N).
 
--spec(upstream_failures([upstream()]) -> non_neg_integer()).
+-spec(upstream_failures(upstream()) -> number()).
```

Comment:
Prometheus library stores all metrics using much different data structure. I'd recommend to re-write `take_upstreams/2` function considering this knowledge, e.g.

```erlang
-spec(take_upstreams([upstream()]) -> [upstream()]).
take_upstreams(Upstreams) when length(Upstreams) < 3 ->
    Upstreams;
take_upstreams(Upstreams) ->
    Buckets = upstream_buckets(Upstreams),
    case lists:unzip(Buckets) of
        {_N, [[_, _ | _] = Bucket | _Buckets]} ->
            choose(2, Bucket);
        {_N, [[Upstream], Bucket | _Buckets]} ->
            [Upstream | choose(1, Bucket)]
    end.

-spec(upstream_failures() -> orddict:orddict(upstream(), pos_integer())).
upstream_failures() ->
    AllFailures = prometheus_counter:values(dns, worker_failures_total),
    Failures = [{Address, Count} || {Labels, Count} <- AllFailures,
                                    {"upstream_address", Address} <- Labels],
    lists:foldl(fun ({Address, Count}, Acc) ->
        orddict:update_counter(Address, Count, Acc)
    end, orddict:new(), Failures).

-spec(upstream_buckets([upstream()]) ->
    orddict:orddict(non_neg_integer(), [upstream()])).
upstream_buckets(Upstreams) ->
    Failures = upstream_failures(),
    lists:foldl(fun (Upstream, Acc) ->
        Address = upstream_to_binary(Upstream),
        Count =
            case orddict:find(Address, Failures) of
                {ok, N} -> N;
                error -> 0
            end,
        orddict:append_list(Count, [Upstream], Acc)
    end, orddict:new(), Upstreams).
```

Reply:
This line is really awesome:
```
    Failures = [{Address, Count} || {Labels, Count} <- AllFailures,
                                    {"upstream_address", Address} <- Labels]
```